importer.go

```diff
@@ -122,28 +122,56 @@ func (i *Importer) readViaCache(imp string) (jsonnet.Contents, error) {
 		return content, nil
 	}
 
-	u, err := addScheme(imp)
+	content, err := i.fetch(imp)
 	if err != nil {
 		return noContent, err
 	}
-	content, err := fetch(u, i.fetcher())
-	if err == nil {
-		i.cache[imp] = content
+
+	i.cache[imp] = content
+	return content, nil
+}
+
+func (i *Importer) fetch(imp string) (jsonnet.Contents, error) {
+	r, err := i.open(imp)
+	if r == nil || err != nil {
+		return noContent, err
 	}
 
-	return content, err
+	defer r.Close()
+	b, err := ioutil.ReadAll(r)
+	if err != nil {
+		return noContent, err
+	}
+
+	return jsonnet.MakeContents(string(b)), nil
 }
 
-func (i *Importer) fetcher() URLFetcher {
-	if i.Fetcher == nil {
-		// Create a http.Client with a transport handling file:// urls
-		// https://golang.org/pkg/net/http/#NewFileTransport
-		t := &http.Transport{}
-		t.RegisterProtocol("file", http.NewFileTransport(http.Dir("/")))
+func (i *Importer) open(imp string) (io.ReadCloser, error) {
+	if !isNetpath(imp) {
+		r, err := os.Open(imp)
+		if os.IsNotExist(err) {
+			return nil, nil
+		}
+		return r, err
+	}
+
+	resp, err := i.fetcher().Get("https:" + imp)
+	if err != nil {
+		return nil, err
+	}
 
-		i.Fetcher = &http.Client{Transport: t}
+	if resp.StatusCode == http.StatusNotFound {
+		return nil, nil
+	} else if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("could not fetch %#v: %s", imp, resp.Status)
 	}
+	return resp.Body, nil
+}
 
+func (i *Importer) fetcher() URLFetcher {
+	if i.Fetcher == nil {
+		i.Fetcher = &http.Client{}
```

Comment:
I don't understand the need for lazy initialization here.
I think we have been here before and looked at http.Client source struct and it was fairly small, but maybe I am confusing this with something else.
IMO lazy initialization is extra cognitive load for the reader (no need for this function at all without it) and should only be used if really justified (ie measurable performance or memory footprint improvement).

Reply:
There's no *need* for lazy initialisation, but the Go proverb "Make the zero value useful" kind of demands it. You can just create `i := jsonnext.Importer{}` and you have a useful value. Not sure how to do that without this lazy initialisation. The alternative is to require a constructor - `NewImporter()`, but I like to be able to initialise just the fields I want on initialisation, which the constructor does not let me do.

However I do not this is not concurrency-safe. I don't think the jsonnet vm is concurrent (although it probably isn't specified that it is not), but perhaps you might want to use the same importer in different vms concurrently. I don't really want to add a `sync.Once` for this. I'll think about it - this is not part of this PR, so it gets to live for now ;-)